Small improvement to x86 vectorized hashes and medium-sizes hash.

goldsteinn opened the pull request.

goldsteinn committed:
Use `vpsrlq` instead of `vpshufd` in avx2/avx512 accumulate.

`vpshufd` was just to essentially implement `dst >> 32` which can (and
should) be done with `vpsrlq`.

`XXH3_{32|64|128}bits` throughput seems to be limitted by a port_5
bottleneck and under-uses port_0.

Hashing 1mb in a loop 100000 times with `vpshufd`:
    x86-64-v4:
         5,363,095,005      cycles
         5,195,740,817      port_0
            17,104,516      port_1
         5,358,811,641      port_5
    x86-64-v3:
         7,954,143,671      cycles
         6,610,562,251      port_0
         7,405,810,291      port_1
         7,474,008,658      port_5

`vpshufd` on x86-64-v3 only runs on port_5 (and for x86-64-v4
only runs on port_5 when used with `zmm` registers). `vpsrlq` can
run on port_{0|1} (port_0 only for `zmm`) which helps to
alleviate this bottleneck.

Hashing 1mb in a loop 100000 times with `vpsrlq`:
    x86-64-v4:
         5,305,723,460      cycles
         5,246,052,997      port_0
            17,759,291      port_1
         5,303,090,711      port_5
    x86-64-v3:
         7,852,139,442      cycles
         6,950,368,526      port_0
         7,301,134,590      port_1
         7,215,987,871      port_5

which in general gets a better port distribution and results in
slightly better times.

Note the sse2 accumulate also would benefit from shifting uops from
port5 -> port0

    x86-64-v2 with `pshufd`:
        13,128,826,786      port_0
        15,014,228,537      port_1
        14,801,410,216      port_5

    x86-64-v2 with `psrlq`:
        13,899,468,407      port_0
        14,774,790,828      port_1
        14,555,300,082      port_5

But since `pshufd` has a 3-term encoding and `psrlq` only has a 2-term
encoding, using `psrlq` causes an extra `movdqa` in codegen which
is a net-negative (at least on Icelake benchmarks).

Benchmarking on Icelake this commit results in the following
improvements:

NB: The avx2 changes where also benchmarked on Icelake which is
not really ideal. Although note that Icelake (unlike native
x86-64-v3 hardware has a faster `vpshufd` (port_{1|5} vs
just port_5) so hopefully anything skewed by incorrect hardware
was infavor of the original implementation).

goldsteinn committed:
Avoid signed truncation in hash 17to128 and 129to240

Signed truncation of len gets worse codegen so keep it unsigned.

Signed codegen for loop bound on x86-64-v4:
```
     mov    %esi,%ecx
     lea    0xe(%rsi),%r10d
     dec    %ecx
     cmovns %ecx,%r10d
     sar    $0x4,%r10d
```

Unsigned codegen for loop bound on x86-64-v4:
```
    lea    -0x1(%rsi),%ecx
    mov    %ecx,%r9d
    shr    $0x4,%r9d
```

goldsteinn committed:
Replace 2x vpxor with vpternlogd in avx512 scramble

Neither Clang nor GCC <= 11.2 get this so seems worth it to do by hand
so save the instruction. Not really on the critical path so only minor
perf changes.

Hashing 8kb On Icelake with 2x vpxor:

     5,451,267,137      cycles
     5,371,545,012      p0
       301,488,579      p1
     5,420,594,430      p5
       450,583,083      p6

Hashing 8kb On Icelake with 2x vpternlogd:

     5,400,583,791      cycles
     5,329,872,021      p0
       305,543,626      p1
     5,372,500,814      p5
       446,853,381      p6

goldsteinn committed:
Speedup size hash for 17to128 by using independent accumulators.

Faster (better IPC) to make computation slightly more independent.

Results are from geometric mean of N = 6 runs.

Machine:
```
11th Gen Intel(R) Core(TM) i7-11850H @ 2.50GHz
```

CC:
```
gcc (Ubuntu 11.3.0-1ubuntu1~22.04) 11.3.0
```

Bench Command:
```
// Change is only for 17-128 range.
$> ./benchHash xxh3 --mins=1 --maxs=256 --minl=0 --maxl=0
```

Aggregated Results for [17, 128] size range:
Times reported as geometric mean of all speedups.

Latency for small inputs of fixed size :
    - 1.073
Throughput small inputs of fixed size (from 1 to 256 bytes):
    - 1.173
Latency for small inputs of random size [1-N] :
    - 1.051
benchmarking random size inputs [1-N] :
    - 1.134

So roughly 5-17% improvement.

goldsteinn commented:
NB: Results for the last change (numbers attached).

## Results are from geometric mean of N = 6 runs.

#### Machine:
```
11th Gen Intel(R) Core(TM) i7-11850H @ 2.50GHz
```

#### CC:
```
gcc (Ubuntu 11.3.0-1ubuntu1~22.04) 11.3.0
```

#### Bench Command:
```
# Change is only for 17-128 range.
$> ./benchHash xxh3 --mins=1 --maxs=256 --minl=0 --maxl=0
```

#### Aggregated Results for [17, 128] size range:

Times reported as geometric mean of all speedups.

Latency for small inputs of fixed size : 
    - 1.073
Throughput small inputs of fixed size (from 1 to 256 bytes): 
    - 1.173
Latency for small inputs of random size [1-N] : 
    - 1.051
benchmarking random size inputs [1-N] : 
    - 1.134

So roughly 5-17% improvement.

Latency for small inputs of fixed size :
Size,TPut-old  ,TPut-new  ,Tput Speedup
17  ,1.320E+08 ,1.402E+08 ,1.062
18  ,1.374E+08 ,1.442E+08 ,1.049
19  ,1.354E+08 ,1.399E+08 ,1.033
20  ,1.319E+08 ,1.376E+08 ,1.043
21  ,1.403E+08 ,1.472E+08 ,1.049
22  ,1.376E+08 ,1.417E+08 ,1.030
23  ,1.353E+08 ,1.405E+08 ,1.038
24  ,1.357E+08 ,1.398E+08 ,1.030
25  ,1.317E+08 ,1.368E+08 ,1.038
26  ,1.331E+08 ,1.360E+08 ,1.021
27  ,1.374E+08 ,1.430E+08 ,1.040
28  ,1.364E+08 ,1.451E+08 ,1.064
29  ,1.313E+08 ,1.370E+08 ,1.043
30  ,1.316E+08 ,1.360E+08 ,1.034
31  ,1.303E+08 ,1.351E+08 ,1.037
32  ,1.312E+08 ,1.380E+08 ,1.052
33  ,1.145E+08 ,1.232E+08 ,1.076
34  ,1.143E+08 ,1.235E+08 ,1.081
35  ,1.139E+08 ,1.220E+08 ,1.071
36  ,1.140E+08 ,1.226E+08 ,1.075
37  ,1.140E+08 ,1.223E+08 ,1.072
38  ,1.148E+08 ,1.234E+08 ,1.075
39  ,1.142E+08 ,1.222E+08 ,1.071
40  ,1.152E+08 ,1.249E+08 ,1.084
41  ,1.113E+08 ,1.181E+08 ,1.061
42  ,1.139E+08 ,1.223E+08 ,1.074
43  ,1.128E+08 ,1.215E+08 ,1.078
44  ,1.148E+08 ,1.215E+08 ,1.058
45  ,1.142E+08 ,1.213E+08 ,1.062
46  ,1.153E+08 ,1.230E+08 ,1.067
47  ,1.131E+08 ,1.202E+08 ,1.063
48  ,1.141E+08 ,1.226E+08 ,1.075
49  ,1.137E+08 ,1.220E+08 ,1.073
50  ,1.137E+08 ,1.217E+08 ,1.070
51  ,1.146E+08 ,1.229E+08 ,1.072
52  ,1.133E+08 ,1.199E+08 ,1.059
53  ,1.127E+08 ,1.197E+08 ,1.061
54  ,1.141E+08 ,1.209E+08 ,1.059
55  ,1.136E+08 ,1.211E+08 ,1.066
56  ,1.139E+08 ,1.214E+08 ,1.066
57  ,1.142E+08 ,1.227E+08 ,1.075
58  ,1.142E+08 ,1.219E+08 ,1.067
59  ,1.125E+08 ,1.194E+08 ,1.061
60  ,1.139E+08 ,1.198E+08 ,1.052
61  ,1.154E+08 ,1.201E+08 ,1.041
62  ,1.123E+08 ,1.198E+08 ,1.067
63  ,1.142E+08 ,1.201E+08 ,1.052
64  ,1.135E+08 ,1.200E+08 ,1.057
65  ,1.021E+08 ,1.091E+08 ,1.069
66  ,1.021E+08 ,1.089E+08 ,1.066
67  ,1.022E+08 ,1.090E+08 ,1.067
68  ,1.007E+08 ,1.086E+08 ,1.078
69  ,1.009E+08 ,1.086E+08 ,1.077
70  ,1.023E+08 ,1.088E+08 ,1.063
71  ,1.014E+08 ,1.084E+08 ,1.070
72  ,1.011E+08 ,1.093E+08 ,1.081
73  ,1.018E+08 ,1.089E+08 ,1.071
74  ,1.010E+08 ,1.078E+08 ,1.067
75  ,1.017E+08 ,1.097E+08 ,1.078
76  ,1.015E+08 ,1.089E+08 ,1.073
77  ,1.018E+08 ,1.090E+08 ,1.071
78  ,1.006E+08 ,1.096E+08 ,1.090
79  ,1.012E+08 ,1.090E+08 ,1.078
80  ,1.030E+08 ,1.097E+08 ,1.064
81  ,1.016E+08 ,1.088E+08 ,1.071
82  ,1.046E+08 ,1.109E+08 ,1.060
83  ,1.021E+08 ,1.089E+08 ,1.066
84  ,1.016E+08 ,1.098E+08 ,1.081
85  ,1.024E+08 ,1.096E+08 ,1.071
86  ,1.015E+08 ,1.087E+08 ,1.071
87  ,1.023E+08 ,1.092E+08 ,1.068
88  ,1.018E+08 ,1.089E+08 ,1.070
89  ,1.007E+08 ,1.082E+08 ,1.075
90  ,1.020E+08 ,1.083E+08 ,1.062
91  ,1.015E+08 ,1.083E+08 ,1.067
92  ,1.022E+08 ,1.093E+08 ,1.070
93  ,1.022E+08 ,1.088E+08 ,1.064
94  ,1.014E+08 ,1.084E+08 ,1.070
95  ,1.028E+08 ,1.078E+08 ,1.049
96  ,1.018E+08 ,1.096E+08 ,1.077
97  ,9.125E+07 ,9.894E+07 ,1.084
98  ,9.028E+07 ,9.916E+07 ,1.098
99  ,9.052E+07 ,1.001E+08 ,1.105
100 ,9.008E+07 ,9.996E+07 ,1.110
101 ,9.017E+07 ,1.001E+08 ,1.110
102 ,9.052E+07 ,9.922E+07 ,1.096
103 ,8.951E+07 ,9.959E+07 ,1.113
104 ,8.992E+07 ,9.945E+07 ,1.106
105 ,9.010E+07 ,1.000E+08 ,1.110
106 ,9.049E+07 ,9.907E+07 ,1.095
107 ,9.056E+07 ,9.899E+07 ,1.093
108 ,9.104E+07 ,9.843E+07 ,1.081
109 ,9.014E+07 ,1.000E+08 ,1.110
110 ,8.961E+07 ,9.872E+07 ,1.102
111 ,9.093E+07 ,9.915E+07 ,1.090
112 ,9.060E+07 ,9.936E+07 ,1.097
113 ,8.968E+07 ,9.909E+07 ,1.105
114 ,9.056E+07 ,9.996E+07 ,1.104
115 ,9.134E+07 ,9.861E+07 ,1.080
116 ,9.101E+07 ,9.970E+07 ,1.095
117 ,9.038E+07 ,9.909E+07 ,1.096
118 ,9.017E+07 ,9.846E+07 ,1.092
119 ,9.009E+07 ,9.907E+07 ,1.100
120 ,9.061E+07 ,9.906E+07 ,1.093
121 ,8.989E+07 ,9.993E+07 ,1.112
122 ,9.059E+07 ,9.917E+07 ,1.095
123 ,9.146E+07 ,9.963E+07 ,1.089
124 ,9.077E+07 ,9.875E+07 ,1.088
125 ,9.035E+07 ,9.919E+07 ,1.098
126 ,9.060E+07 ,9.953E+07 ,1.099
127 ,9.047E+07 ,9.889E+07 ,1.093
128 ,9.023E+07 ,9.934E+07 ,1.101

Throughput small inputs of fixed size (from 1 to 256 bytes):
Size,TPut-old  ,TPut-new  ,Tput Speedup
17  ,3.200E+08 ,3.771E+08 ,1.178
18  ,3.190E+08 ,3.756E+08 ,1.177
19  ,3.193E+08 ,3.736E+08 ,1.170
20  ,3.205E+08 ,3.754E+08 ,1.171
21  ,3.177E+08 ,3.752E+08 ,1.181
22  ,3.192E+08 ,3.764E+08 ,1.179
23  ,3.199E+08 ,3.757E+08 ,1.175
24  ,3.204E+08 ,3.749E+08 ,1.170
25  ,3.179E+08 ,3.749E+08 ,1.179
26  ,3.205E+08 ,3.750E+08 ,1.170
27  ,3.188E+08 ,3.760E+08 ,1.179
28  ,3.197E+08 ,3.754E+08 ,1.174
29  ,3.208E+08 ,3.744E+08 ,1.167
30  ,3.179E+08 ,3.756E+08 ,1.182
31  ,3.206E+08 ,3.740E+08 ,1.167
32  ,3.182E+08 ,3.741E+08 ,1.176
33  ,2.208E+08 ,2.612E+08 ,1.183
34  ,2.199E+08 ,2.622E+08 ,1.193
35  ,2.202E+08 ,2.643E+08 ,1.201
36  ,2.194E+08 ,2.643E+08 ,1.205
37  ,2.193E+08 ,2.628E+08 ,1.198
38  ,2.192E+08 ,2.612E+08 ,1.192
39  ,2.210E+08 ,2.611E+08 ,1.182
40  ,2.214E+08 ,2.624E+08 ,1.185
41  ,2.205E+08 ,2.629E+08 ,1.192
42  ,2.199E+08 ,2.621E+08 ,1.192
43  ,2.211E+08 ,2.614E+08 ,1.182
44  ,2.207E+08 ,2.608E+08 ,1.182
45  ,2.225E+08 ,2.610E+08 ,1.173
46  ,2.217E+08 ,2.620E+08 ,1.182
47  ,2.204E+08 ,2.635E+08 ,1.195
48  ,2.214E+08 ,2.638E+08 ,1.191
49  ,2.224E+08 ,2.626E+08 ,1.181
50  ,2.206E+08 ,2.608E+08 ,1.183
51  ,2.228E+08 ,2.607E+08 ,1.170
52  ,2.215E+08 ,2.616E+08 ,1.181
53  ,2.206E+08 ,2.623E+08 ,1.189
54  ,2.214E+08 ,2.614E+08 ,1.180
55  ,2.209E+08 ,2.601E+08 ,1.178
56  ,2.205E+08 ,2.599E+08 ,1.179
57  ,2.233E+08 ,2.600E+08 ,1.164
58  ,2.200E+08 ,2.622E+08 ,1.192
59  ,2.206E+08 ,2.623E+08 ,1.189
60  ,2.219E+08 ,2.632E+08 ,1.186
61  ,2.226E+08 ,2.622E+08 ,1.178
62  ,2.224E+08 ,2.613E+08 ,1.175
63  ,2.199E+08 ,2.614E+08 ,1.189
64  ,2.218E+08 ,2.629E+08 ,1.186
65  ,1.733E+08 ,1.971E+08 ,1.137
66  ,1.739E+08 ,1.964E+08 ,1.129
67  ,1.731E+08 ,1.966E+08 ,1.135
68  ,1.733E+08 ,1.973E+08 ,1.139
69  ,1.737E+08 ,1.968E+08 ,1.133
70  ,1.734E+08 ,1.977E+08 ,1.140
71  ,1.730E+08 ,1.966E+08 ,1.137
72  ,1.743E+08 ,2.038E+08 ,1.169
73  ,1.729E+08 ,1.951E+08 ,1.129
74  ,1.724E+08 ,1.967E+08 ,1.141
75  ,1.724E+08 ,1.969E+08 ,1.143
76  ,1.706E+08 ,1.970E+08 ,1.154
77  ,1.717E+08 ,1.970E+08 ,1.147
78  ,1.721E+08 ,1.956E+08 ,1.137
79  ,1.724E+08 ,1.957E+08 ,1.135
80  ,1.748E+08 ,2.011E+08 ,1.151
81  ,1.714E+08 ,1.983E+08 ,1.157
82  ,1.718E+08 ,1.980E+08 ,1.153
83  ,1.715E+08 ,1.982E+08 ,1.156
84  ,1.704E+08 ,1.985E+08 ,1.165
85  ,1.701E+08 ,1.969E+08 ,1.157
86  ,1.697E+08 ,1.979E+08 ,1.167
87  ,1.710E+08 ,1.974E+08 ,1.154
88  ,1.737E+08 ,2.039E+08 ,1.174
89  ,1.702E+08 ,1.984E+08 ,1.166
90  ,1.692E+08 ,1.984E+08 ,1.173
91  ,1.703E+08 ,1.978E+08 ,1.162
92  ,1.712E+08 ,1.973E+08 ,1.152
93  ,1.715E+08 ,1.964E+08 ,1.145
94  ,1.714E+08 ,1.976E+08 ,1.153
95  ,1.709E+08 ,1.972E+08 ,1.154
96  ,1.739E+08 ,2.025E+08 ,1.165
97  ,1.409E+08 ,1.643E+08 ,1.166
98  ,1.396E+08 ,1.641E+08 ,1.175
99  ,1.394E+08 ,1.643E+08 ,1.179
100 ,1.393E+08 ,1.635E+08 ,1.174
101 ,1.401E+08 ,1.642E+08 ,1.172
102 ,1.400E+08 ,1.635E+08 ,1.168
103 ,1.405E+08 ,1.640E+08 ,1.167
104 ,1.420E+08 ,1.671E+08 ,1.176
105 ,1.400E+08 ,1.645E+08 ,1.176
106 ,1.400E+08 ,1.634E+08 ,1.167
107 ,1.394E+08 ,1.644E+08 ,1.179
108 ,1.399E+08 ,1.644E+08 ,1.175
109 ,1.388E+08 ,1.638E+08 ,1.180
110 ,1.401E+08 ,1.639E+08 ,1.170
111 ,1.405E+08 ,1.647E+08 ,1.172
112 ,1.434E+08 ,1.665E+08 ,1.162
113 ,1.383E+08 ,1.654E+08 ,1.196
114 ,1.375E+08 ,1.644E+08 ,1.196
115 ,1.386E+08 ,1.653E+08 ,1.193
116 ,1.391E+08 ,1.651E+08 ,1.187
117 ,1.384E+08 ,1.649E+08 ,1.191
118 ,1.385E+08 ,1.654E+08 ,1.194
119 ,1.376E+08 ,1.650E+08 ,1.199
120 ,1.417E+08 ,1.672E+08 ,1.180
121 ,1.377E+08 ,1.647E+08 ,1.196
122 ,1.375E+08 ,1.652E+08 ,1.202
123 ,1.381E+08 ,1.648E+08 ,1.193
124 ,1.375E+08 ,1.660E+08 ,1.207
125 ,1.380E+08 ,1.647E+08 ,1.193
126 ,1.372E+08 ,1.659E+08 ,1.209
127 ,1.377E+08 ,1.659E+08 ,1.205
128 ,1.413E+08 ,1.672E+08 ,1.184

Latency for small inputs of random size [1-N] :
Size,TPut-old  ,TPut-new  ,Tput Speedup
17  ,1.470E+08 ,1.492E+08 ,1.015
18  ,1.468E+08 ,1.490E+08 ,1.015
19  ,1.456E+08 ,1.482E+08 ,1.018
20  ,1.454E+08 ,1.478E+08 ,1.017
21  ,1.454E+08 ,1.481E+08 ,1.019
22  ,1.442E+08 ,1.466E+08 ,1.016
23  ,1.431E+08 ,1.470E+08 ,1.027
24  ,1.423E+08 ,1.456E+08 ,1.023
25  ,1.437E+08 ,1.467E+08 ,1.021
26  ,1.423E+08 ,1.459E+08 ,1.025
27  ,1.408E+08 ,1.449E+08 ,1.029
28  ,1.411E+08 ,1.451E+08 ,1.028
29  ,1.417E+08 ,1.455E+08 ,1.027
30  ,1.413E+08 ,1.443E+08 ,1.021
31  ,1.410E+08 ,1.450E+08 ,1.029
32  ,1.400E+08 ,1.440E+08 ,1.029
33  ,1.387E+08 ,1.444E+08 ,1.041
34  ,1.390E+08 ,1.428E+08 ,1.028
35  ,1.381E+08 ,1.429E+08 ,1.035
36  ,1.362E+08 ,1.406E+08 ,1.032
37  ,1.367E+08 ,1.419E+08 ,1.038
38  ,1.357E+08 ,1.402E+08 ,1.033
39  ,1.342E+08 ,1.398E+08 ,1.042
40  ,1.345E+08 ,1.395E+08 ,1.037
41  ,1.335E+08 ,1.388E+08 ,1.040
42  ,1.335E+08 ,1.387E+08 ,1.039
43  ,1.325E+08 ,1.380E+08 ,1.042
44  ,1.311E+08 ,1.364E+08 ,1.041
45  ,1.317E+08 ,1.377E+08 ,1.046
46  ,1.304E+08 ,1.366E+08 ,1.048
47  ,1.296E+08 ,1.356E+08 ,1.046
48  ,1.304E+08 ,1.368E+08 ,1.049
49  ,1.291E+08 ,1.350E+08 ,1.046
50  ,1.293E+08 ,1.354E+08 ,1.047
51  ,1.287E+08 ,1.353E+08 ,1.051
52  ,1.295E+08 ,1.349E+08 ,1.042
53  ,1.284E+08 ,1.349E+08 ,1.051
54  ,1.275E+08 ,1.341E+08 ,1.052
55  ,1.277E+08 ,1.342E+08 ,1.050
56  ,1.278E+08 ,1.344E+08 ,1.052
57  ,1.278E+08 ,1.342E+08 ,1.050
58  ,1.270E+08 ,1.333E+08 ,1.049
59  ,1.257E+08 ,1.322E+08 ,1.051
60  ,1.261E+08 ,1.332E+08 ,1.056
61  ,1.258E+08 ,1.327E+08 ,1.055
62  ,1.250E+08 ,1.319E+08 ,1.055
63  ,1.264E+08 ,1.325E+08 ,1.048
64  ,1.256E+08 ,1.321E+08 ,1.052
65  ,1.245E+08 ,1.318E+08 ,1.059
66  ,1.247E+08 ,1.320E+08 ,1.059
67  ,1.241E+08 ,1.309E+08 ,1.054
68  ,1.245E+08 ,1.302E+08 ,1.046
69  ,1.233E+08 ,1.301E+08 ,1.055
70  ,1.222E+08 ,1.296E+08 ,1.060
71  ,1.228E+08 ,1.289E+08 ,1.050
72  ,1.214E+08 ,1.282E+08 ,1.056
73  ,1.218E+08 ,1.284E+08 ,1.054
74  ,1.220E+08 ,1.288E+08 ,1.056
75  ,1.203E+08 ,1.277E+08 ,1.061
76  ,1.213E+08 ,1.280E+08 ,1.055
77  ,1.204E+08 ,1.276E+08 ,1.060
78  ,1.205E+08 ,1.275E+08 ,1.058
79  ,1.214E+08 ,1.278E+08 ,1.053
80  ,1.190E+08 ,1.256E+08 ,1.055
81  ,1.187E+08 ,1.255E+08 ,1.058
82  ,1.192E+08 ,1.262E+08 ,1.059
83  ,1.195E+08 ,1.262E+08 ,1.056
84  ,1.185E+08 ,1.260E+08 ,1.064
85  ,1.195E+08 ,1.267E+08 ,1.060
86  ,1.187E+08 ,1.257E+08 ,1.059
87  ,1.188E+08 ,1.260E+08 ,1.061
88  ,1.184E+08 ,1.257E+08 ,1.062
89  ,1.166E+08 ,1.242E+08 ,1.065
90  ,1.188E+08 ,1.261E+08 ,1.062
91  ,1.177E+08 ,1.250E+08 ,1.063
92  ,1.166E+08 ,1.230E+08 ,1.055
93  ,1.168E+08 ,1.239E+08 ,1.060
94  ,1.160E+08 ,1.228E+08 ,1.059
95  ,1.166E+08 ,1.235E+08 ,1.060
96  ,1.167E+08 ,1.234E+08 ,1.057
97  ,1.168E+08 ,1.240E+08 ,1.061
98  ,1.167E+08 ,1.236E+08 ,1.059
99  ,1.166E+08 ,1.236E+08 ,1.060
100 ,1.155E+08 ,1.226E+08 ,1.062
101 ,1.156E+08 ,1.227E+08 ,1.061
102 ,1.144E+08 ,1.215E+08 ,1.062
103 ,1.146E+08 ,1.216E+08 ,1.061
104 ,1.132E+08 ,1.201E+08 ,1.061
105 ,1.140E+08 ,1.213E+08 ,1.064
106 ,1.135E+08 ,1.205E+08 ,1.062
107 ,1.140E+08 ,1.215E+08 ,1.066
108 ,1.128E+08 ,1.200E+08 ,1.064
109 ,1.123E+08 ,1.193E+08 ,1.063
110 ,1.115E+08 ,1.192E+08 ,1.069
111 ,1.116E+08 ,1.192E+08 ,1.068
112 ,1.115E+08 ,1.191E+08 ,1.068
113 ,1.115E+08 ,1.190E+08 ,1.068
114 ,1.115E+08 ,1.191E+08 ,1.068
115 ,1.116E+08 ,1.189E+08 ,1.066
116 ,1.105E+08 ,1.186E+08 ,1.074
117 ,1.106E+08 ,1.177E+08 ,1.065
118 ,1.109E+08 ,1.184E+08 ,1.068
119 ,1.106E+08 ,1.182E+08 ,1.069
120 ,1.096E+08 ,1.173E+08 ,1.070
121 ,1.097E+08 ,1.174E+08 ,1.071
122 ,1.088E+08 ,1.171E+08 ,1.076
123 ,1.105E+08 ,1.180E+08 ,1.068
124 ,1.093E+08 ,1.168E+08 ,1.068
125 ,1.077E+08 ,1.164E+08 ,1.081
126 ,1.094E+08 ,1.171E+08 ,1.070
127 ,1.082E+08 ,1.162E+08 ,1.074
128 ,1.100E+08 ,1.180E+08 ,1.073

benchmarking random size inputs [1-N] :
Size,TPut-old  ,TPut-new  ,Tput Speedup
17  ,4.010E+08 ,4.385E+08 ,1.094
18  ,3.900E+08 ,4.389E+08 ,1.125
19  ,3.879E+08 ,4.367E+08 ,1.126
20  ,3.814E+08 ,4.348E+08 ,1.140
21  ,3.814E+08 ,4.282E+08 ,1.122
22  ,3.762E+08 ,4.210E+08 ,1.119
23  ,3.760E+08 ,4.192E+08 ,1.115
24  ,3.736E+08 ,4.200E+08 ,1.124
25  ,3.705E+08 ,4.220E+08 ,1.139
26  ,3.683E+08 ,4.199E+08 ,1.140
27  ,3.676E+08 ,4.172E+08 ,1.135
28  ,3.644E+08 ,4.124E+08 ,1.132
29  ,3.655E+08 ,4.111E+08 ,1.125
30  ,3.637E+08 ,4.104E+08 ,1.128
31  ,3.623E+08 ,4.109E+08 ,1.134
32  ,3.583E+08 ,4.081E+08 ,1.139
33  ,3.544E+08 ,4.051E+08 ,1.143
34  ,3.478E+08 ,3.960E+08 ,1.139
35  ,3.421E+08 ,3.931E+08 ,1.149
36  ,3.337E+08 ,3.815E+08 ,1.143
37  ,3.361E+08 ,3.818E+08 ,1.136
38  ,3.342E+08 ,3.793E+08 ,1.135
39  ,3.200E+08 ,3.688E+08 ,1.153
40  ,3.239E+08 ,3.701E+08 ,1.143
41  ,3.150E+08 ,3.603E+08 ,1.144
42  ,3.189E+08 ,3.642E+08 ,1.142
43  ,3.140E+08 ,3.558E+08 ,1.133
44  ,3.038E+08 ,3.489E+08 ,1.149
45  ,3.093E+08 ,3.538E+08 ,1.144
46  ,2.988E+08 ,3.444E+08 ,1.152
47  ,2.993E+08 ,3.407E+08 ,1.138
48  ,2.998E+08 ,3.440E+08 ,1.147
49  ,2.954E+08 ,3.383E+08 ,1.145
50  ,2.911E+08 ,3.379E+08 ,1.161
51  ,2.903E+08 ,3.327E+08 ,1.146
52  ,2.862E+08 ,3.347E+08 ,1.169
53  ,2.926E+08 ,3.356E+08 ,1.147
54  ,2.860E+08 ,3.289E+08 ,1.150
55  ,2.885E+08 ,3.299E+08 ,1.143
56  ,2.898E+08 ,3.301E+08 ,1.139
57  ,2.856E+08 ,3.281E+08 ,1.149
58  ,2.790E+08 ,3.225E+08 ,1.156
59  ,2.755E+08 ,3.173E+08 ,1.152
60  ,2.801E+08 ,3.237E+08 ,1.155
61  ,2.742E+08 ,3.186E+08 ,1.162
62  ,2.749E+08 ,3.153E+08 ,1.147
63  ,2.772E+08 ,3.200E+08 ,1.155
64  ,2.774E+08 ,3.171E+08 ,1.143
65  ,2.663E+08 ,3.103E+08 ,1.165
66  ,2.695E+08 ,3.095E+08 ,1.149
67  ,2.637E+08 ,3.064E+08 ,1.162
68  ,2.686E+08 ,3.097E+08 ,1.153
69  ,2.586E+08 ,2.991E+08 ,1.157
70  ,2.577E+08 ,2.962E+08 ,1.149
71  ,2.596E+08 ,2.966E+08 ,1.142
72  ,2.503E+08 ,2.888E+08 ,1.154
73  ,2.546E+08 ,2.938E+08 ,1.154
74  ,2.561E+08 ,2.933E+08 ,1.145
75  ,2.493E+08 ,2.844E+08 ,1.141
76  ,2.531E+08 ,2.883E+08 ,1.139
77  ,2.462E+08 ,2.812E+08 ,1.142
78  ,2.487E+08 ,2.837E+08 ,1.141
79  ,2.504E+08 ,2.861E+08 ,1.143
80  ,2.395E+08 ,2.766E+08 ,1.155
81  ,2.420E+08 ,2.739E+08 ,1.132
82  ,2.437E+08 ,2.767E+08 ,1.136
83  ,2.402E+08 ,2.747E+08 ,1.143
84  ,2.397E+08 ,2.719E+08 ,1.134
85  ,2.416E+08 ,2.773E+08 ,1.148
86  ,2.380E+08 ,2.702E+08 ,1.135
87  ,2.397E+08 ,2.754E+08 ,1.149
88  ,2.380E+08 ,2.705E+08 ,1.136
89  ,2.318E+08 ,2.619E+08 ,1.130
90  ,2.370E+08 ,2.708E+08 ,1.142
91  ,2.364E+08 ,2.680E+08 ,1.134
92  ,2.307E+08 ,2.642E+08 ,1.145
93  ,2.292E+08 ,2.623E+08 ,1.145
94  ,2.288E+08 ,2.589E+08 ,1.131
95  ,2.301E+08 ,2.614E+08 ,1.136
96  ,2.307E+08 ,2.633E+08 ,1.141
97  ,2.317E+08 ,2.612E+08 ,1.128
98  ,2.266E+08 ,2.552E+08 ,1.126
99  ,2.278E+08 ,2.567E+08 ,1.127
100 ,2.241E+08 ,2.483E+08 ,1.108
101 ,2.258E+08 ,2.496E+08 ,1.106
102 ,2.204E+08 ,2.419E+08 ,1.098
103 ,2.207E+08 ,2.432E+08 ,1.102
104 ,2.142E+08 ,2.352E+08 ,1.098
105 ,2.203E+08 ,2.462E+08 ,1.118
106 ,2.162E+08 ,2.388E+08 ,1.105
107 ,2.184E+08 ,2.399E+08 ,1.099
108 ,2.148E+08 ,2.366E+08 ,1.101
109 ,2.090E+08 ,2.304E+08 ,1.102
110 ,2.082E+08 ,2.290E+08 ,1.100
111 ,2.067E+08 ,2.319E+08 ,1.122
112 ,2.068E+08 ,2.316E+08 ,1.120
113 ,2.084E+08 ,2.348E+08 ,1.127
114 ,2.093E+08 ,2.346E+08 ,1.121
115 ,2.080E+08 ,2.352E+08 ,1.130
116 ,2.064E+08 ,2.339E+08 ,1.133
117 ,2.033E+08 ,2.280E+08 ,1.122
118 ,2.049E+08 ,2.297E+08 ,1.121
119 ,2.054E+08 ,2.284E+08 ,1.112
120 ,2.015E+08 ,2.245E+08 ,1.114
121 ,2.032E+08 ,2.288E+08 ,1.126
122 ,1.991E+08 ,2.253E+08 ,1.131
123 ,2.047E+08 ,2.298E+08 ,1.123
124 ,1.988E+08 ,2.247E+08 ,1.130
125 ,1.966E+08 ,2.219E+08 ,1.129
126 ,1.975E+08 ,2.238E+08 ,1.133
127 ,1.969E+08 ,2.220E+08 ,1.128
128 ,2.009E+08 ,2.285E+08 ,1.137
[results.tar.gz](https://github.com/Cyan4973/xxHash/files/9968245/results.tar.gz)

Cyan4973 commented on xxhash.h:
```diff
@@ -4002,8 +4004,8 @@ XXH3_len_129to240_64b(const xxh_u8* XXH_RESTRICT input, size_t len,
     {   xxh_u64 acc = len * XXH_PRIME64_1;
-        int const nbRounds = (int)len / 16;
-        int i;
+        unsigned int const nbRounds = (unsigned int)len / 16;
```
minor : any reason you prefer to employ `unsigned` instead of `int` for these variables ?

goldsteinn replied:
Yeah, [unsigned seems to get better codegen](https://github.com/Cyan4973/xxHash/pull/754/commits/1456dc5cbf51ec810c44a7a17ed0faf8d17a3eaa) for GCC11/12.

goldsteinn replied:
> Thanks @goldsteinn ! I can definitely confirm a notable bandwidth gains for the 17-128 variant of `xxh3` running on `x64` and compiled with `gcc`. The latency gain is much smaller, but it's also not a loss, so that's fine.
> 
> I would like to test on `arm64` | `m1` to confirm that the benefit is generic.

Great! Let me know if need to make it arch specific (although expect this helps any OOE machine).

Cyan4973 replied:
Yeah, likely the `/ 16` operation can only be safely translated into `>> 4` on `unsigned` types.

Cyan4973 commented:
Thanks @goldsteinn !
I can definitely confirm a notable bandwidth gains for the 17-128 variant of `xxh3` running on `x64` and compiled with `gcc`.
The latency gain is much smaller, but it's also not a loss, so that's fine.

I would like to test on `arm64` | `m1` to confirm that the benefit is generic.

goldsteinn commented:
> Thanks @goldsteinn ! I can definitely confirm a notable bandwidth gains for the 17-128 variant of `xxh3` running on `x64` and compiled with `gcc`. The latency gain is much smaller, but it's also not a loss, so that's fine.
> 
> I would like to test on `arm64` | `m1` to confirm that the benefit is generic.

Great! Let me know if need to make it arch specific (although expect this helps any OOE machine).

goldsteinn commented:
NB: Numbers for the most recently commit (changing how loop boundaries are done for 129to240 regions).
Changes are pretty minimal but its pretty obviously inoptimal codegen.

Edit: platform / bench command same as above.
[results-64.tar.gz](https://github.com/Cyan4973/xxHash/files/9994080/results-64.tar.gz)
[results-64-summary.txt](https://github.com/Cyan4973/xxHash/files/9994081/results-64-summary.txt)
[results-128.tar.gz](https://github.com/Cyan4973/xxHash/files/9994082/results-128.tar.gz)
[results-128-summary.txt](https://github.com/Cyan4973/xxHash/files/9994083/results-128-summary.txt)

goldsteinn commented:
@Cyan4973 any chance you would reconsider supporting `XXH_UNSTABLE` on master branch? I have a fork [xxhash-unstable](https://github.com/goldsteinn/xxHash-Unstable) but commits like this last one (the one updating the loop bounds for 129to240 hashes) are a bit of a pain as they will essentially need to be duplicated to not using `i <= (len - {16|32})` in a new patch. Future changes that touch any of the functions with the erroneous inclusive upper bound will have the same issue.

Also intuitively there are many other uses like myself that use this for hashtables / non-persistent applications and can benefit from the performance gains.

t-mat commented:
@goldsteinn As for `__builtin_unreachable()`, perhaps it's a good time to introduce `XXH_UNREACHABLE()` since we'll be able to use it as a part of standard C23.

It may look like this

```C
#if defined(__STDC_VERSION__) && (__STDC_VERSION__ > 201710L)
/* C23 and future versions have standard "unreachable()" */
#  include <stddef.h>
#  define XXH_UNREACHABLE() unreachable()

#elif defined(__GNUC__) && (__GNUC__ >= 5)
#  define XXH_UNREACHABLE() __builtin_unreachable()

#elif defined(_MSC_VER)
#  define XXH_UNREACHABLE() __assume(0)

#else
#  define XXH_UNREACHABLE()
#endif
```

edit : Capitalize

goldsteinn commented:
> @goldsteinn As for `__builtin_unreachable()`, perhaps it's a good time to introduce `XXH_UNREACHABLE()` since we'll be able to use it as a part of standard C23.
> 
> It may look like this
> 
> ```c
> #if defined(__STDC_VERSION__) && (__STDC_VERSION__ > 201710L)
> /* C23 and future versions have standard "unreachable()" */
> #  include <stddef.h>
> #  define XXH_UNREACHABLE() unreachable()
> 
> #elif defined(__GNUC__) && (__GNUC__ >= 5)
> #  define XXH_UNREACHABLE() __builtin_unreachable()
> 
> #elif defined(_MSC_VER)
> #  define XXH_UNREACHABLE() __assume(0)
> 
> #else
> #  define XXH_UNREACHABLE()
> #endif
> ```
> 
> edit : Capitalize



> @goldsteinn As for `__builtin_unreachable()`, perhaps it's a good time to introduce `XXH_UNREACHABLE()` since we'll be able to use it as a part of standard C23.
> 
> It may look like this
> 
> ```c
> #if defined(__STDC_VERSION__) && (__STDC_VERSION__ > 201710L)
> /* C23 and future versions have standard "unreachable()" */
> #  include <stddef.h>
> #  define XXH_UNREACHABLE() unreachable()
> 
> #elif defined(__GNUC__) && (__GNUC__ >= 5)
> #  define XXH_UNREACHABLE() __builtin_unreachable()
> 
> #elif defined(_MSC_VER)
> #  define XXH_UNREACHABLE() __assume(0)
> 
> #else
> #  define XXH_UNREACHABLE()
> #endif
> ```
> 
> edit : Capitalize

@t-mat 
Makes sense. Do we need to seperate version check for `clang` or does `__GNUC__ >= 5` work?

Edit:
For the `__builtin_unreachable` case it may be easier to just `XXH_HAS_BUILTIN(__builtin_unreachable)`.

t-mat commented:
> For the __builtin_unreachable case it may be easier to just XXH_HAS_BUILTIN(__builtin_unreachable).

Nice :+1:

----

I realized we also need definition for C++23.  It may similar to the definition of `XXH_ALIGN`

```C
...
#elif defined(__cplusplus) && (__cplusplus > 202002L)
/* C++23 and future versions have std::unreachable() */
#  include <utility> /* std::unreachable() */
#  define XXH_UNREACHABLE() std::unreachable()
#else
...
```

goldsteinn commented:
> ```c
> #if defined(__STDC_VERSION__) && (__STDC_VERSION__ > 201710L)
> /* C23 and future versions have standard "unreachable()" */
> #  include <stddef.h>
> #  define XXH_UNREACHABLE() unreachable()
> 
> #elif defined(__GNUC__) && (__GNUC__ >= 5)
> #  define XXH_UNREACHABLE() __builtin_unreachable()
> 
> #elif defined(_MSC_VER)
> #  define XXH_UNREACHABLE() __assume(0)
> 
> #else
> #  define XXH_UNREACHABLE()
> #endif
> ```

Thanks :)

Suprising changes codegen a fair bit changing i/len to `unsigned` + using `XXH_ASSUME` (not sure if for the better), will run some benchmarks before posting.

t-mat commented:
As you know, C++23, clang and MSVC have `[[assume(cond)]];`, `__builtin_assume(cond);` and `__assume(cond);`.

But unfortunately C23 and gcc don't have `assume`.  We may be able to roughly emulate it by `if(!(cond)) { XXH_UNREACHABLE(); }` though.

goldsteinn commented:
> As you know, C++23, clang and MSVC have `[[assume(cond)]];`, `__builtin_assume(cond);` and `__assume(cond);`.
> 
> But unfortunately C23 and gcc don't have `assume`. We may be able to roughly emulate it by `if(!(cond)) { XXH_UNREACHABLE(); }` though.

That's what I did. Pushed updated version (new codegen makes sense, had inverted the condition in `XXH_ASSUME` when it was wacky).

t-mat commented on xxhash.h:
```diff
@@ -2028,6 +2028,28 @@ static int XXH_isLittleEndian(void)
+#elif defined(__cplusplus) && (__cplusplus > 202002L)
+/* C++23 and future versions have std::unreachable() */
+#  include <utility> /* std::unreachable() */
+#  define XXH_UNREACHABLE() std::unreachable();
```
This is my mistake.  But we should remove the last `;`.  Sorry for my inaccurate code!

goldsteinn replied:
good catch. Fixed!

goldsteinn committed:
Fix '-Wsign-conversion' warning in XXH3_initCustomSecret_avx512

No change to generated code, just suppresses the warning when using
stricter compiler checks.

goldsteinn commented:
> Thanks @goldsteinn ! I can definitely confirm a notable bandwidth gains for the 17-128 variant of `xxh3` running on `x64` and compiled with `gcc`. The latency gain is much smaller, but it's also not a loss, so that's fine.
> 
> I would like to test on `arm64` | `m1` to confirm that the benefit is generic.

Any updates?
@Cyan4973

Cyan4973 commented:
> Any updates?

Hi @goldsteinn ,

I've just been trying your PR on a M1 system,
and it immediately hits this compilation issue : 

```
../../xxhash.h:4062:9: error: expected a for, while, or do-while loop to follow '#pragma clang loop'
        acc_end = 0;
        ^
```

Not sure if it's M1 specific, it rather looks like a compiler error.
If it's more generally `macos` related, then I'm surprised it doesn't hit any problem in CI...

The local compiler used : 
`Apple clang version 14.0.0 (clang-1400.0.29.202)`

_Edit_:
GH has a `macos` test for `xxHash`. It works fine there:
https://github.com/Cyan4973/xxHash/actions/runs/3459035922/jobs/5774022012
The main difference is that it uses `macos-11` and `clang-13`, 
while the M1 test system uses `macos-13` and `clang-14`.

_Edit 2_:
To be fair, the solution is simple : just move the `acc_end = 0;` statement above the `#ifdef` block where the `#pragma` is set.

Cyan4973 commented:
OK, I've been testing this PR on `macos` with `M1 Pro`, with a small modification to source code (as described above). The results are a bit surprising.

There is small bandwidth gain in the 17-128 range, but it's minimal, much smaller than the one observed on `x64`.
OK, but that's still a gain, and since it's not a loss, well it seems okay.

Except that ... the surprise is that the 129-240 range is negatively impacted, heavily. We are talking ~-20%  bandwidth loss, which is quite significant.
This was unexpected, and so far, I have not been able to identify a clear reason nor a fix.
Note that the code for `129-240` is updated by this PR, so there is a definitely a relation.
I tried to update or remove the `#pragma clang loop` in front of the second loop, but it did not change the outcome. Apparently, this loop can't be vectorized anyway.
Some `x64` can be prone to "instruction alignment" side effects (skylane-era cpus notably), resulting in random performance swings, but I was expecting `arm64` cpus to be more or less immune to this effect. Unfortunately, while `x64` cpus can offer rich counters for performance introspection, I know no equivalent for M1, so it's difficult to analyze.
A more probable cause is that the code change might break some compiler optimization for 129-240, or something like that.

On the latency front, there is no major difference between the PR and `dev`. This remains true at all ranges, including 16-128 and 129-240.

So all the focus seems on bandwidth tests.

goldsteinn commented:
> OK, I've been testing this PR on `macos` with `M1 Pro`, with a small modification to source code (as described above). The results are a bit surprising.
> 
> There is small bandwidth gain in the 17-128 range, but it's minimal, much smaller than the one observed on `x64`. OK, but that's still a gain, and since it's not a loss, well it seems okay.
> 
> Except that ... the surprise is that the 129-240 range is negatively impacted, heavily. We are talking ~-20% bandwidth loss, which is quite significant. This was unexpected, and so far, I have not been able to identify a clear reason nor a fix. Note that the code for `129-240` is updated by this PR, so there is a definitely a relation. I tried to update or remove the `#pragma clang loop` in front of the second loop, but it did not change the outcome. Apparently, this loop can't be vectorized anyway. Some `x64` can be prone to "instruction alignment" side effects (skylane-era cpus notably), resulting in random performance swings, but I was expecting `arm64` cpus to be more or less immune to this effect. Unfortunately, while `x64` cpus can offer rich counters for performance introspection, I know no equivalent for M1, so it's difficult to analyze. A more probable cause is that the code change might break some compiler optimization for 129-240, or something like that.
> 

Does the 20% regression stay consistent if you change the relative alignment of 129_240? Ime maybe with alignment % 64 in {0, 8, 16,..}? (Don't have m1 to test on unfortunately.

> On the latency front, there is no major difference between the PR and `dev`. This remains true at all ranges, including 16-128 and 129-240.
> 
> So all the focus seems on bandwidth tests.

How do you want to proceed? Remove the 129-240 changes from the series or figure out what's going on first?

Cyan4973 commented:
I will have difficulties to find time for an investigation on 129-240 range for M1 on my side,
and it seems you can't reproduce the test and investigate on your side.

In which case, it seems the simpler follow up is to revert the changes to the 129-240 range.

If I understand correctly, only the 17-128 range is supposed to benefit from this PR,
so there's little point in modifying the 129-240 range anyway.

goldsteinn commented:
> I will have difficulties to find time for an investigation on 129-240 range for M1 on my side, and it seems you can't reproduce the test and investigate on your side.
> 
> In which case, it seems the simpler follow up is to revert the changes to the 129-240 range.
> 
> If I understand correctly, only the 17-128 range is supposed to benefit from this PR, so there's little point in modifying the 129-240 range anyway.

Removed the patch changing `nbRounds` in 129-240 range.

One thing I thought of, is maybe this broke so over/underflow check that enabled some optimizations.

If you have the time the following patch might fix it:

```
diff --git a/xxhash.h b/xxhash.h
index 525985c..8daff01 100644
--- a/xxhash.h
+++ b/xxhash.h
@@ -1803,7 +1803,7 @@ static void* XXH_memcpy(void* dest, const void* src, size_t size)
 #  include <assert.h>   /* note: can still be disabled with NDEBUG */
 #  define XXH_ASSERT(c)   assert(c)
 #else
-#  define XXH_ASSERT(c)   ((void)0)
+#  define XXH_ASSERT(c)   XXH_ASSUME(c)
 #endif
 
 /* note: use after variable declarations */
```

goldsteinn commented:
> > I will have difficulties to find time for an investigation on 129-240 range for M1 on my side, and it seems you can't reproduce the test and investigate on your side.
> > In which case, it seems the simpler follow up is to revert the changes to the 129-240 range.
> > If I understand correctly, only the 17-128 range is supposed to benefit from this PR, so there's little point in modifying the 129-240 range anyway.
> 
> Removed the patch changing `nbRounds` in 129-240 range.

Didn't remove the patch changing signed -> unsigned casts which also touches `129to240`. LMK
if you want that dropped as well.
> 
> One thing I thought of, is maybe this broke so over/underflow check that enabled some optimizations.
> 
> If you have the time the following patch might fix it:
> 
> ```
> diff --git a/xxhash.h b/xxhash.h
> index 525985c..8daff01 100644
> --- a/xxhash.h
> +++ b/xxhash.h
> @@ -1803,7 +1803,7 @@ static void* XXH_memcpy(void* dest, const void* src, size_t size)
>  #  include <assert.h>   /* note: can still be disabled with NDEBUG */
>  #  define XXH_ASSERT(c)   assert(c)
>  #else
> -#  define XXH_ASSERT(c)   ((void)0)
> +#  define XXH_ASSERT(c)   XXH_ASSUME(c)
>  #endif
>  
>  /* note: use after variable declarations */
> ```

Cyan4973 commented:
Yep, this solves the problem for the 129-240 range, while preserving the bandwidth benefit for 17-128 one.
Hence it looks good to me!

goldsteinn commented:
> Yep, this solves the problem for the 129-240 range, while preserving the bandwidth benefit for 17-128 one. Hence it looks good to me!

This being the revert or using asume when assert is disabled?

Cyan4973 commented:
> > Yep, this solves the problem for the 129-240 range, while preserving the bandwidth benefit for 17-128 one. Hence it looks good to me!
> 
> This being the revert or using asume when assert is disabled?

The current code is this PR, aka the revert.

Not sure how I could try the other suggestion, since it seems to require modifying [`e630cbc`](https://github.com/Cyan4973/xxHash/commit/e630cbcccb7395170d3744ae826a69c26e0f5968) which is no longer part of the `git` history.

goldsteinn commented:
> > > Yep, this solves the problem for the 129-240 range, while preserving the bandwidth benefit for 17-128 one. Hence it looks good to me!
> > 
> > 
> > This being the revert or using asume when assert is disabled?
> 
> The current code is this PR, aka the revert.
> 
> Not sure how I could try the other suggestion, since it seems to require modifying [`e630cbc`](https://github.com/Cyan4973/xxHash/commit/e630cbcccb7395170d3744ae826a69c26e0f5968) which is no longer part of the `git` history.

After this goes in I'll make a new PR for it w/ the `assume` logic.

Btw any chance you reconsidered the value of `XXH_UNSTABLE`?

Cyan4973 commented:
> Btw any chance you reconsidered the value of `XXH_UNSTABLE`?

If you mean hosting a variant that produces different hash values,
then I'm afraid I'm too concerned at this point by the risk of confusion for the xxHash ecosystem.
